### PR TITLE
fixes for Quantum-Espresso 6.5 with MKL and ELPA

### PR DESCRIPTION
**Pull request by tz-rrze**
- qe-6.5 fails to detect MKL if BLAS_LIBS is set
- allow compilation with ELPA>2016

successfully tested with ComposerXE2020, Intel-MPI & OpenMPI-3.1.5, and elpa@2019.05.002 on CentOS-7.7/x86_64

### CODE REVIEW

**Comment by tz-rrze** [comment]
1) The configure script of QE already had some magic to react on changes in the ELPA API; however, the configure script only supported ELPA up to 2016. The patch "qe-elpa-configure.patch" simply mimics the existing logic for more recent version and is from scratch.
The upstream QE code seems to still only contain a partial fix for newer ELPA versions (cf. lines 8396-8405 vs. 8416-8424 of https://gitlab.com/QEF/q-e/-/blob/develop/install/configure)

2) The 2nd patch (qe-elpa.patch-v1) is required because the 2019 ELPA release has removed the "legacy" API (use elpa1/use elpa2) while a few minor details changed already between 2016 and 2019.
This patch is based on https://gitlab.com/ionuorah1/q-e/commit/fcfd0874c5f5bdb66f132f86c4e194611282769a with parts of the PW/src/setup.f90 changes removed as they already exist in the staged QE-6.5 sources in SPACK.

**Comment by adamjstewart** [comment]
@naromero77

**Comment by tz-rrze** [comment]
I just updated my own patch file to not only patch the provided configure script but also the underlying m4 script so that the fix is also included in case people rerun autoconf on the SPACK provided source tree; would not be necessary for the normal SPACK use case but that's saver now.

**Comment by prckent** [comment]
@tz-rrze Have you tried to get these changes upstreamed at https://gitlab.com/QEF/q-e ?

**Comment by tz-rrze** [comment]
@prckent I did not yet try myself to upstream the changes; moreover, 
- one change is already pending with the QE developers since 5 months (https://gitlab.com/QEF/q-e/issues/144),
- the other one (MKL not detected if BLAS_LIBS is set) originates only from the way SPACK is calling configure (I'd therefore assume a won't fix from the q-e developers)
- for the 3rd one (configure not being fully aware of ELPA>2016) I have to verify the issue with a vanilla q-e first (but then again it depends on q-e issue 144 to be applied); so far I only did it within SPACK.

But of course, I understand that the local patches should be kept minimal with changes being upstreamed whenever possible.

**Comment by alalazo** [comment]
> the other one (MKL not detected if BLAS_LIBS is set) originates only from the way SPACK is calling configure (I'd therefore assume a won't fix from the q-e developers)

Can you elaborate on this? Wondering if we can fix it without a patch.

**Comment by tz-rrze** [comment]
@alalazo the "MKL not detected" does not require any of the patch files but is handled in changed var/spack/repos/builtin/packages/quantum-espresso/package.py; cf. changes of line 233 to 240-242 in connection with the new lines 222-225, i.e. do not set BLAS_LIBS for @6.5^mkl

**Comment by alalazo** [comment]
@tz-rrze Apologies, since you were discussing changes to be upstreamed I gave it for granted that they were patches.

**Comment by tz-rrze** [comment]
@alalazo no problem
If you and the others do not like the patches for "allow compilation with ELPA>2016" one would have to modify the PR to at least make the quantum-espresso/package.py conflict with any ELPA > 2016.05.004. Right now, without the patches, building quantum-espresso will fail with ^elpa as the preferred version is 2019.05.002.
The q-e developers did dislike the patch for recent ELPA in their https://gitlab.com/QEF/q-e/issues/144 but are discussion if it should be merged directly or through a complementary European project (MaX) - if I understand it correctly. Unfortunately, they merged a small part of the change already so that just pulling the patch from them no longer works. The change on configure/m4 is only cosmetics to get one define correctly into the Makefile.

**Comment by tz-rrze** [comment]
> @prckent I did not yet try myself to upstream the changes; moreover,
> 
> * one change is already pending with the QE developers since 5 months (https://gitlab.com/QEF/q-e/issues/144),
> * the other one (MKL not detected if BLAS_LIBS is set) originates only from the way SPACK is calling configure (I'd therefore assume a won't fix from the q-e developers)
> * for the 3rd one (configure not being fully aware of ELPA>2016) I have to verify the issue with a vanilla q-e first (but then again it depends on q-e issue 144 to be applied); so far I only did it within SPACK.
> 
> But of course, I understand that the local patches should be kept minimal with changes being upstreamed whenever possible.

Upstream accepted PRs for the 1st and 3rd item:
- https://gitlab.com/QEF/q-e/-/commit/20c9141a104dc6e5259cef73c331d5dee991df8e / https://gitlab.com/QEF/q-e/-/merge_requests/808
- https://gitlab.com/QEF/q-e/-/commit/5f706710686b859ada0f961fafbb22f827bf6ed6 / 
https://gitlab.com/QEF/q-e/-/merge_requests/809

**Comment by alalazo** [comment]
@tz-rrze That's great. My opinion then would be to avoid patches and just merge the second item. The other fixes will come with upstream on the next point release. Waiting to hear @naromero77

**Comment by naromero77** [comment]
Sorry, I have been most most of the week and have not been able to keep up with e-mail. Let me take a look.

**Comment by naromero77** [comment]
There are frequently things that are broken in QE's configure. I have spent a lot of time patching stuff with filter_files -- it is hard to keep up with them in general and hard to "draw the line" w.r.t. to what should get fixed now and how versus just waiting for the next release.

I only patch the build system for the QE release with upstream QE commits whenever this is possible. I also only patch the build system and nothing else, e.g. QE science features that are broken should probably just stay broken until the next release. Generally, we should strive to get these fixes/patches into upstream QE.

Instead of patching QE 6.5 with your own patch, are you able to backport these two commits:
https://gitlab.com/QEF/q-e/-/commit/20c9141a104dc6e5259cef73c331d5dee991df8e / 
https://gitlab.com/QEF/q-e/-/commit/5f706710686b859ada0f961fafbb22f827bf6ed6 /
and apply them to QE 6.5 release? Previously in this thread, it looks like you attempted to do this but there was a problem due to `setup.F90`?

Here are some examples of how I do this:
https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/quantum-espresso/package.py#L153-L170

I also have a few comments about this PR.

File: var/spack/repos/builtin/packages/quantum-espresso/package.py [hunk 1]
```diff
@@ -75,6 +75,9 @@ class QuantumEspresso(Package):
     patch('dspev_drv_elpa.patch', when='@6.1.0:+patch+elpa ^elpa@2016.05.004')
     patch('dspev_drv_elpa.patch', when='@6.1.0:+patch+elpa ^elpa@2016.05.003')
 
+    patch('qe-elpa.patch-v1', when='@6.5^elpa@2018:2019')
+    patch('qe-elpa-configure.patch', when='@6.5^elpa@2017:2019')
```
Comment (naromero77) [inline var/spack/repos/builtin/packages/quantum-espresso/package.py]: Please include a `+patch` to these when statements.

File: var/spack/repos/builtin/packages/quantum-espresso/package.py [hunk 2]
```diff
@@ -216,6 +219,10 @@ def install(self, spec, prefix):
             options.append(
                 'FFTW_INCLUDE={0}'.format(join_path(env['MKLROOT'],
                                                     'include/fftw')))
+            options.append(
+                'LIBDIRS={0}'.format(join_path(env['MKLROOT'],
+                                     'lib/intel64')))
```
Comment (naromero77) [inline var/spack/repos/builtin/packages/quantum-espresso/package.py]: @adamjstewart Isn't there away to add `lib/intel64` in an architecture agnostic manner?

Comment (tz-rrze) [reply var/spack/repos/builtin/packages/quantum-espresso/package.py]: `self.spec['mkl'].prefix.lib` probably should do; will check that

**Comment by tz-rrze** [comment]
> Instead of patching QE 6.5 with your own patch, are you able to backport these two commits:
> https://gitlab.com/QEF/q-e/-/commit/20c9141a104dc6e5259cef73c331d5dee991df8e

This one cannot directly be applied as it only patches the _autoconf/m4_ file. (If we do not patch `configure` directly we would have to run _autoconf_ - which seems to be a bad or even worse idea.)

> https://gitlab.com/QEF/q-e/-/commit/5f706710686b859ada0f961fafbb22f827bf6ed6 /
> and apply them to QE 6.5 release? Previously in this thread, it looks like you attempted to do this but there was a problem due to `setup.F90`?

This one should apply without further modifications as it does not include the part of the original patch from https://gitlab.com/QEF/q-e/issues/144 which has already been integrated to `setup.F90` and caused the conflict.

**Comment by tz-rrze** [comment]
I probably should split that PR into two separate ones (fix MKL detection which is SPACK-only // ELPA feature upgrade) ...

**Comment by alalazo** [comment]
> I probably should split that PR into two separate ones (fix MKL detection which is SPACK-only // ELPA feature upgrade) ...

If you do, please keep this one for the patches PR so the discussion is not lost.

File: var/spack/repos/builtin/packages/quantum-espresso/package.py (hunk 2 above)
Comment (tz-rrze) [reply var/spack/repos/builtin/packages/quantum-espresso/package.py]: this change even can be omitted completely :-)

**Comment by adamjstewart** [comment]
What's the status of this PR? It looks like it needs to be rebased.

**Comment by alalazo** [comment]
Closing as stale

**Pull request closed without merging**